### ISSUE TITLE
Support for camera images

### ISSUE THREAD

**Issue by joa77** [tomer8007/kik-bot-api-unofficial#130]
I love to see, that sending images is now supported. Will there be a way of sending images as camera images?

**Comment by bluemods** [comment]
The ```<k``` tag needs to be fixed by @tomer8007 first otherwise even if you send a camera message it will be changed to gallery upon receiving.

**Comment by tomer8007** [comment]
@bluemods If you know exactly how to get it right you are more than welcome to submit a PR.

**Comment by bluemods** [comment]
> @bluemods If you know exactly how to get it right you are more than welcome to submit a PR.

I'm very inexperienced in python, I coded it in Java

**Comment by MisterWorld** [comment]
@bluemods Your python code is very good. Can you help to solve this problem?

**Comment by tomer8007** [comment]
@bluemods The `signed, lang, sid, ts, v, cv, conn, dev, n` generation code is already here at [EstablishAuthenticatedSessionRequest](https://github.com/tomer8007/kik-bot-api-unofficial/blob/5c63963835961091d331da2e44141ba89baf2a9e/kik_unofficial/datatypes/xmpp/login.py#L76), right?

So all you're saying is we need to replace `anon=""` with `anon="1"` and use this code even when `self.username is None and self.password is None` [here](https://github.com/tomer8007/kik-bot-api-unofficial/blob/2a791a9012aa3197c84559bc4be243ee09a83401/kik_unofficial/client.py#L78)?

**Comment by bluemods** [comment]
Correct. It still won't pass the initial verification because the map sorting is incorrect, but it's a good start.

the ```dev``` is NOT in the authentication ```<k``` stanza, only in the anonymous ```<k``` stanza. 

Key differences between the anon and auth <k:

-The ```dev``` attribute is your Device ID
-The create hmac attribute (```cv```) uses CAN + your Device ID in place of your JID, since we aren't supposed to know that yet.
-The ```signed``` attribute *also* uses= CAN + your Device ID in place of your JID to make the RSA sign, since we aren't supposed to know that yet.

-The ```from``` attribute is omitted from the anonymous <k tag
-The ```to``` attribute is also omitted from the anonymous <k tag
-The ```p``` attribute is also omitted from the anonymous <k tag since we aren't logging in yet

That's about it I believe.

**Comment by tomer8007** [comment]
OK, sounds sensible. Why is the sorting incorrect? Isn't the code at [sort_kik_map](https://github.com/tomer8007/kik-bot-api-unofficial/blob/5c63963835961091d331da2e44141ba89baf2a9e/kik_unofficial/utilities/cryptographic_utilities.py#L134) correct?

Also why leaving the currect state would prevent sending camera images? The authentication stanza is working right now.

**Comment by bluemods** [comment]
Well try it for yourself, change ```com.kik.ext.gallery``` to ```com.kik.ext.camera```, you'll see Kik change it right back when you receive it.

The best way to beat a hacker is make them think they've won, especially so you don't perfect your responses because that way they really can't detect the bots. So they're liberal with what they let sign on, but there are plenty of shadow punishments such as frequent captchas and group joining restrictions. They don't want you to realize that the client is busted and is getting marked as a bot sign in.

**Comment by tomer8007** [comment]
I agree we can't send camera images, the question is why would changing the anonymous stanza fix it (did you try it?).

**Comment by tomer8007** [comment]
Also the sorting of the map is fine, right?

**Comment by bluemods** [comment]
> Also the sorting of the map is fine, right?

I don't know, but I put some fake values to show you what it should convert to.

Notice how there's a bunch of whitespace before the actual <k? (26 to be exact) that's because it's based on a hashcode of the map, and your client ignores this completely. If you can plus in these fake values into your map sorter and have it print out the exact same result, you'll be good.

```                          <k signed="bar" conn="foo" n="foo" cv="foo" ts="foo" from="foo" sid="foo" p="foo" to="foo" v="foo" lang="foo">```

**Comment by bluemods** [comment]
Here's a raw pastebin for further clarification on what you're missing: https://pastebin.com/qzwhi9DY

**Comment by bluemods** [comment]
Well I'm impressed, the python API spits it out 100% correct, it is just missing the leading whitespaces.

**Comment by tomer8007** [comment]
So finally, what needs to be changed so that `com.kik.ext.camera` will work?

**Comment by bluemods** [comment]
You need to append the leading whitespaces which is based off a hashcode of the sorted map from earlier. They are checked as part of the security

**Comment by MisterWorld** [comment]
@tomer8007 Hi! How's this features coming?

**Comment by bluemods** [comment]
I have the fix, I could commit it, the problem is is that the bot makers watch this page and it will further spread the problem of the bots abusing it to try and trick people to visit their site or possibly pay money. For that reason it's honestly best this bot remains detectable and camera images do not work

**Comment by joa77** [comment]
I never had a bot sending me images 😅

**Comment by bluemods** [comment]
Go into the anon chats, you'll find one quickly, and then after it's added to this API you'll find many more. Just not a good idea

**Comment by Artificiangel** [comment]
A while back I found that those bots already send fake camera images.
I also noticed some of them don't, probably different devs.

I don't think they (or most) use this API. They were able to mark your messages as delivered before that feature was added here.

**Comment by Artificiangel** [comment]
> -The `dev` attribute is your Device ID and is omitted from the auth <k tag

I had been looking at the decompiled source of the Kik apk (v15.21.0.22201) and converting the parts needed for the sorted <k> tag into python.
All that works, my code outputs the same tag as the java function, but it doesn't let the app use camera images.

I noticed the Kik code has CAN + deviceId assigned to dev. This appears in both the smali and java versions.
I made sure to look at other versions of Kik from multiple sources to make sure. 

If I remove CAN from the dev attribute, Kik returns an error telling me to update my client.

I don't understand what's causing that. 
I am also sending the latest version number and classes dex hash which didn't help.

**Comment by bluemods** [comment]
> > -The `dev` attribute is your Device ID and is omitted from the auth <k tag
> 
> I had been looking at the decompiled source of the Kik apk (v15.21.0.22201) and converting the parts needed for the sorted tag into python.
> All that works, my code outputs the same tag as the java function, but it doesn't let the app use camera images.
> 
> I noticed the Kik code has CAN + deviceId assigned to dev. This appears in both the smali and java versions.
> I made sure to look at other versions of Kik from multiple sources to make sure.
> 
> If I remove CAN from the dev attribute, Kik returns an error telling me to update my client.
> 
> I don't understand what's causing that.
> I am also sending the latest version number and classes dex hash which didn't help.

Don't remove the CAN portion of the device ID, it's important. Read this document here for a thorough explanation starting at line 56: https://github.com/tomer8007/kik-bot-api-unofficial/blob/new/kik_unofficial/protobuf/protobuf_source/common_model.proto

Funny story about the historic spammers: there was an exploit back in the day to brute force and make thousands of attempts per second back in 2016 with the old version (now shut down) Kik has always been a dumpster fire.

**Comment by MisterWorld** [comment]
> I have the fix, I could commit it, 

@bluemods @bluemods
 It would be great if you could do it

**Comment by MisterWorld** [comment]
@tomer8007 @bluemods  

Is there any hope for this features?

**Comment by karate21** [comment]
@tomer8007 @bluemods @MisterWorld 
Following up on this. Has there been a solution?
